blog: add Godot XR Progress report Jan 2023

BastiaanOlij opened the pull request:
Progress report for Godot XR

BastiaanOlij committed:
Godot XR Progress report Jan 2023

YuriSizov commented:
Can you make the cover image with a wide aspect? We would also probably need to standardize it through our thumbnail generator: https://godot-image-generator.vercel.app/thumbnail

BastiaanOlij commented:
> Can you make the cover image with a wide aspect? We would also probably need to standardize it through our thumbnail generator: https://godot-image-generator.vercel.app/thumbnail

Thats always an issue in VR with screenshots being tall, I'll see if I can crop it, what is a good size for it?

YuriSizov commented:
1280x720 seems to be the norm for our articles these days.

BastiaanOlij commented:
Ok image is updated, fixed one spelling mistake and added a small blurp about the XR editor.

Calinou commented on collections/_article/godot-xr-progress-update-jan-2023.md:
```diff
@@ -0,0 +1,79 @@
+date: 2023-01-28 00:00:00
+---
+
+An update of the state of XR in Godot is long overdue, so lets look at some highlights of 2022 and what we're working on now that it is 2023.
```
```suggestion
An update of the state of XR in Godot is long overdue. Let's look at some highlights of 2022 and what we're working on now that it is 2023.
```

Calinou commented on collections/_article/godot-xr-progress-update-jan-2023.md:
```diff
@@ -0,0 +1,79 @@
+
+Godot XR Tools is our XR toolkit that offers a number of implementations of popular XR mechanics.
+
+Recently we released version 3.2.0 of this toolkit, this version has a number of quality of life improvements, some nice new movement providers and adds support for WebXR into the toolkit.
```
```suggestion
Godot XR Tools is our XR toolkit that offers a number of implementations of popular XR mechanics. Recently, we released version 3.2.0 of this toolkit. This version has a number of quality of life improvements, some nice new movement providers and adds support for WebXR into the toolkit.
```

Calinou commented on collections/_article/godot-xr-progress-update-jan-2023.md:
```diff
@@ -0,0 +1,79 @@
+
+Recently we released version 3.2.0 of this toolkit, this version has a number of quality of life improvements, some nice new movement providers and adds support for WebXR into the toolkit.
+
+We also revamped the documentation of the toolkit which is now hosted at https://godotvr.github.io/godot-xr-tools/
```
```suggestion
We also revamped the documentation of the toolkit which is now hosted at [godotvr.github.io/godot-xr-tools](https://godotvr.github.io/godot-xr-tools/).
```

Calinou commented on collections/_article/godot-xr-progress-update-jan-2023.md:
```diff
@@ -0,0 +1,79 @@
+
+Contributor Teddybear082 has used Godot XR Tools to implement VR ports of various open source projects showing how easy it is to apply our toolkit. His latest project he is working on is a port of the highly successful Cruelty Squad.
+
+Godot XR Tools can be found on Godot's asset library.
```
```suggestion
While the stable release of XR Tools remains a Godot 3.x plugin, work is nearly completed on the Godot 4 port. Most of the plugin now works in Godot 4.

For the near future, we're working on a template project for Godot that has a whole XR setup ready to go.
We're also working on a new player controller that should make it easier to drop XR support into an existing first-person game.

Contributor Teddybear082 has used Godot XR Tools to implement VR ports of various open source projects. This shows how easy it is to apply our toolkit. His latest project he is working on is a port of the highly successful [Cruelty Squad](https://godotengine.org/showcase/cruelty-squad/).

Godot XR Tools can be found on the [asset library](https://godotengine.org/asset-library/asset/214).
```

Calinou commented on collections/_article/godot-xr-progress-update-jan-2023.md:
```diff
@@ -0,0 +1,79 @@
+
+### Godot documentation sprint
+
+Godot organised a documentation sprint this January to ensure the documentation for Godot 4 was brought up to spec in anticipation of Godot 4s stable release. XR of course was part of this sprint. Various documentation pages were completed and are now under review. Others are still being worked on. Expect to see the pages added to the documentation soon.
```
```suggestion
[Engine developers have organized a documentation sprint this January](https://godotengine.org/article/godot-4-0-docs-sprint/) to ensure the documentation for Godot 4 was brought up to spec in anticipation of Godot 4's stable release. XR of course was part of this sprint. Various documentation pages were completed and are now under review; others are still being worked on. Expect to see new pages added to the documentation soon.
```

Calinou commented on collections/_article/godot-xr-progress-update-jan-2023.md:
```diff
@@ -0,0 +1,79 @@
+
+For low end devices with limited Vulkan support, OpenGL was added as a backend to Godot 4. This implementation of OpenGL is based on OpenGL ES 3.0 but its architecture is more friendly to low end/mobile devices.
+
+While there are a few loose ends, contributor David Snopek has added stereoscopic rendering support to the OpenGL renderer and it has proven to be a very efficient solution for XR. This is an upgrade for anyone using the GLES2 renderer in Godot 3 to consider. This implementation takes advantage of the improvements in OpenGL ES 3 but does not make use of the techniques the Godot 3 GLES3 renderer a ...
```
```suggestion
For low-end devices with limited Vulkan support, [OpenGL was added as a backend to Godot 4](https://godotengine.org/article/status-of-opengl-renderer/). This implementation of OpenGL is based on OpenGL ES 3.0, but its architecture is more friendly to low-end/mobile devices compared to Godot 3's GLES3 backend.

While there are a few loose ends, contributor David Snopek has added stereoscopic rendering support to the OpenGL renderer. This has proven to be a very efficient solution for XR. This is an upgrade for anyone using the GLES2 renderer in Godot 3 to consider. This implementation takes advantage of the improvements in OpenGL ES 3, but does *not* make use of Godot 3 GLES3 renderer technniques that were detrimental to performance on mobile XR. The only footnote is that the implementation *requires* the OpenGL multiview extension.
```

Calinou commented on collections/_article/godot-xr-progress-update-jan-2023.md:
```diff
@@ -0,0 +1,79 @@
+
+The only footnote is that WebXR has fixed actions bound to controller inputs which puts restrictions on the setup of actions on the OpenXR side to ensure portability between the platforms.
+
+For a future version of Godot 4 we are planning to add WebXR support to the action map implementation as well.
```
```suggestion
The added OpenGL support has enabled us to port our WebXR implementation to Godot 4 as well, again thanks to David Snopek.

Great care was taken to make WebXR support as portable as possible. This allows games developed with Godot to be deployed to both OpenXR and WebXR-based platforms.

The only footnote is that WebXR has fixed actions bound to controller inputs. This puts restrictions on the setup of actions on the OpenXR side to ensure portability between the platforms.

For a future version of Godot 4, we're planning to add WebXR support to the action map implementation as well.
```

Calinou commented on collections/_article/godot-xr-progress-update-jan-2023.md:
```diff
@@ -0,0 +1,79 @@
+
+Support for PICO headsets was recently added to Godot 4, mostly to the credit of Dirk Steinmetz with help from several people at Bytedance. While we're still waiting for a driver update in order to fully support Vulkan, the OpenGL renderer is working fine.
+
+We hope to soon share instructions for setting up Godot on your PICO headset.
```
```suggestion
Support for [PICO headsets](https://www.picoxr.com/) was recently added to Godot 4, mostly to the credit of Dirk Steinmetz with help from several people at Bytedance. While we're still waiting for a driver update to fully support Vulkan, the OpenGL renderer is working fine.

We hope to soon share instructions for setting up Godot on your PICO headset.
```

Calinou commented on collections/_article/godot-xr-progress-update-jan-2023.md:
```diff
@@ -0,0 +1,79 @@
+
+This is the first HMD we are supporting through the official Khronos loader and in theory this should provide support for several other headsets as well.
+
+Again we hope to soon share instructions for setting up Godot on your Magic Leap 2 headset.
```
```suggestion
Support for the [Magic Leap 2](https://www.magicleap.com/magic-leap-2) is in its final stages of development as well, the credit here goes to Ron Bessems.

This is the first HMD we're supporting through the official Khronos loader. In theory, this should provide support for several other headsets as well.

Again, we hope to share instructions soon for setting up Godot on your Magic Leap 2 headset.
```

Calinou commented on collections/_article/godot-xr-progress-update-jan-2023.md:
```diff
@@ -0,0 +1,79 @@
+
+#### Honerary mention
+
+Godot 4 is getting a DirectX backend, this backend doesn't yet have support for stereoscopic rendering but adding this is on the roadmap and with that we're planning to add DirectX support to our OpenXR implementation as well. This will mean native WMR support for PCVR.
```
It's postponed to 4.1: https://github.com/godotengine/godot/pull/70315#issuecomment-1403613525

```suggestion
Godot 4.1 will get a Direct3D 12 backend. This backend doesn't support stereoscopic rendering yet, but adding this is on the roadmap. With that, we're planning to add Direct3D support to our OpenXR implementation as well. This will mean having native Windows Mixed Reality support for PCVR.
```

Calinou commented on collections/_article/godot-xr-progress-update-jan-2023.md:
```diff
@@ -0,0 +1,79 @@
+
+Basic setup is now completed and work is starting on navigating and interacting with the loaded scene. A build that runs natively on the Quest is now also working.
+
+However it must be noted that our focus and attention remains on the release of Godot 4.0 and the inevitable work that comes out of the community feedback. 
```
```suggestion
2022 kept us far busier than expected to get Godot 4.0 to a [state we can release it in](https://godotengine.org/article/release-management-4-0-and-beyond/). As such, work on the XR editor was pushed back. A start has now been made on this interesting feature.

Noteworthy to mention here as well is the work that was done on getting the Godot editor to run on an Android device. This way, game development can be performed on tablets and similar devices. Much of this work is instrumental for getting the VR editor to work. In fact people have already successfully tried running this version of the editor on the Meta Quest!

Basic setup is now completed and work is starting on navigating and interacting with the loaded scene. A build that runs natively on the Meta Quest is now also working.

However, it must be noted that our focus and attention remains on the release of Godot 4.0 and the inevitable work that comes out of the community feedback. 
```

Calinou reviewed (commented):
Should be good to merge after applying suggestions.

YuriSizov commented:
> Should be good to merge after applying suggestions.

Should probably merge on Tuesday or Thursday.

BastiaanOlij commented:
Ok, took a few tries to get all of the changes in, thanks @Calinou for the edits, they were great!

I added a link to the Godot XR template as well now that the first version is in. I also changed the paragraph on Teddybear so it didn't have a double His/he in there. Flows a bit better that way.

akien-mga commented on collections/_article/godot-xr-progress-update-jan-2023.md:
```diff
@@ -0,0 +1,77 @@
+image: /storage/blog/covers/xr-progress-update-redorahnvr3.jpg
+image_caption_title: Redorahn VR
+image_caption_description: VR port of Redorahn, by Teddybear082
+date: 2023-01-30 00:00:00
```
Needs to be updated to 2023-01-31.

akien-mga commented on collections/_article/godot-xr-progress-update-jan-2023.md:
```diff
@@ -0,0 +1,77 @@
+
+Again, we hope to share instructions soon for setting up Godot on your Magic Leap 2 headset.
+
+#### Honerary mention
```
*honorary

akien-mga commented on collections/_article/godot-xr-progress-update-jan-2023.md:
```diff
@@ -0,0 +1,77 @@
+For the near future, we're working on [a template project for Godot](https://github.com/GodotVR/godot-xr-template) that has a whole XR setup ready to go.
+We're also working on a new player controller that should make it easier to drop XR support into an existing first-person game.
+
+Contributor Teddybear082 has used Godot XR Tools to implement VR ports of various open source projects. This shows how easy it is to apply our toolkit. The latest project he is working on is a port of the highly successful [Cruelty Squad](https://godotengine.org/showcase/cruelty-squad/).
```
Would add a link to https://twitter.com/Flat2VR/status/1617699586155638784

akien-mga commented:
Thanks!
